panther_docking/src/panther_charging_dock.cpp

```diff
@@ -0,0 +1,420 @@
+  node_->get_parameter(
+    name + ".external_detection_translation_y", external_detection_translation_y_);
+  node_->get_parameter(
+    name + ".external_detection_translation_z", external_detection_translation_z_);
+
+  double yaw, pitch, roll;
+  node_->get_parameter(name + ".external_detection_rotation_yaw", yaw);
+  node_->get_parameter(name + ".external_detection_rotation_pitch", pitch);
+  node_->get_parameter(name + ".external_detection_rotation_roll", roll);
+
+  external_detection_rotation_.setRPY(roll, pitch, yaw);
+  node_->get_parameter(name + ".docking_distance_threshold", docking_distance_threshold_);
+  node_->get_parameter(name + ".docking_yaw_threshold", docking_yaw_threshold_);
+  node_->get_parameter(name + ".staging_x_offset", staging_x_offset_);
+  node_->get_parameter(name + ".staging_yaw_offset", staging_yaw_offset_);
+
+  node_->get_parameter(
+    name + ".enable_charger_service_call_timeout", enable_charger_service_call_timeout_);
+
+  double filter_coef;
+  node_->get_parameter(name + ".filter_coef", filter_coef);
+  filter_ = std::make_unique<opennav_docking::PoseFilter>(filter_coef, external_detection_timeout_);
+}
+
+void PantherChargingDock::cleanup()
+{
+  dock_pose_pub_.reset();
+  staging_pose_pub_.reset();
+
+  if (panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
+    charging_status_sub_.reset();
+    io_state_sub_.reset();
+    charger_enable_client_.reset();
+  }
+}
+
+void PantherChargingDock::activate()
+
+{
+  dock_pose_pub_ = node_->create_publisher<geometry_msgs::msg::PoseStamped>("docking/dock_pose", 1);
+  staging_pose_pub_ = node_->create_publisher<geometry_msgs::msg::PoseStamped>(
+    "docking/staging_pose", 1);
+
+  if (panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
+    charging_status_sub_ = node_->create_subscription<panther_msgs::msg::ChargingStatus>(
+      "battery/charging_status", rclcpp::SystemDefaultsQoS(),
+      [this](const std::shared_ptr<panther_msgs::msg::ChargingStatus> msg) -> void {
+        charging_status_box_.set(std::move(msg));
+      });
+    io_state_sub_ = node_->create_subscription<panther_msgs::msg::IOState>(
+      "hardware/io_state", rclcpp::SystemDefaultsQoS(),
+      [this](const std::shared_ptr<panther_msgs::msg::IOState> msg) -> void {
+        io_state_box_.set(std::move(msg));
+      });
+
+    charger_enable_client_ =
+      node_->create_client<std_srvs::srv::SetBool>("hardware/charger_enable");
+
+    using namespace std::chrono_literals;
+    if (!charger_enable_client_->wait_for_service(1s)) {
+      RCLCPP_WARN_STREAM(
+        logger_,
+        "Service \"hardware/charger_enable\" not available. Make sure if Panther ROS 2 Stack is "
+        "running.");
+    }
+  }
+}
+
+void PantherChargingDock::deactivate()
+{
+  dock_pose_pub_.reset();
+  staging_pose_pub_.reset();
+
+  if (panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
```

Comment:
A lot of a Panther version checking
- I know that you have an old version for testing but what's the approach of supporting it in general?
-  I support is needed we should rather implement a sort of strategy behavior rather then "ifing" that often